stdlib/ballerina-builtin/src/main/java/org/ballerinalang/nativeimpl/runtime/AuthenticationContext.java

```diff
@@ -0,0 +1,138 @@
+
+import java.util.LinkedHashMap;
+import java.util.Map;
+
+/**
+ * AuthenticationContext represents and holds the authenticated user information.
+ *
+ * @since 0.970.0
+ */
+public class AuthenticationContext {
+
+    public static final int USER_ID_STRING_FIELD_INDEX = 0;
+    public static final int USER_NAME_STRING_FIELD_INDEX = 1;
+    public static final int GROUPS_REF_FIELD_INDEX = 0;
+    public static final int CLAIMS_REF_FIELD_INDEX = 1;
+    public static final int SCOPES_REF_FIELD_INDEX = 2;
+    public static final int AUTH_TYPE_STRING_FIELD_INDEX = 2;
+    public static final int AUTH_TOKEN_STRING_FIELD_INDEX = 3;
+
+    private BStruct authContextStruct;
+
+    public AuthenticationContext(BStruct authContextStruct) {
+        this.authContextStruct = authContextStruct;
+    }
+
+    public String getUserId() {
+        return authContextStruct.getStringField(USER_ID_STRING_FIELD_INDEX);
+    }
+
+    public void setUserId(String userId) {
+        authContextStruct.setStringField(USER_ID_STRING_FIELD_INDEX, userId);
+    }
+
+    public String getUsername() {
+        return authContextStruct.getStringField(USER_NAME_STRING_FIELD_INDEX);
+    }
+
+    public void setUsername(String username) {
+        authContextStruct.setStringField(USER_NAME_STRING_FIELD_INDEX, username);
+    }
+
+    public String[] getGroups() {
+        return getStringArrayField(authContextStruct.getRefField(GROUPS_REF_FIELD_INDEX));
+    }
+
+    public void setGroups(String[] groups) {
+        authContextStruct.setRefField(GROUPS_REF_FIELD_INDEX, new BStringArray(groups));
+    }
+
+    public Map<String, String> getClaims() {
+        return getMapField(authContextStruct.getRefField(CLAIMS_REF_FIELD_INDEX));
+    }
+
+    public void setClaims(Map<String, String> claims) {
+        BMap<String, BString> bMap = new BMap<>();
+        if (claims != null) {
+            claims.forEach((key, value) -> bMap.put(key, new BString(value)));
+        }
+        authContextStruct.setRefField(CLAIMS_REF_FIELD_INDEX, bMap);
+    }
+
+    public String[] getScopes() {
+        return getStringArrayField(authContextStruct.getRefField(SCOPES_REF_FIELD_INDEX));
+
+    }
+
+    public void setScopes(String[] scopes) {
+        authContextStruct.setRefField(SCOPES_REF_FIELD_INDEX, new BStringArray(scopes));
+    }
+
+    public String getAuthType() {
+        return authContextStruct.getStringField(AUTH_TYPE_STRING_FIELD_INDEX);
+    }
+
+    public void setAuthType(String authType) {
+        authContextStruct.setStringField(AUTH_TYPE_STRING_FIELD_INDEX, authType);
+    }
+
+    public String getAuthToken() {
+        return authContextStruct.getStringField(AUTH_TOKEN_STRING_FIELD_INDEX);
+    }
+
+    public void setAuthToken(String authToken) {
+        authContextStruct.setStringField(AUTH_TOKEN_STRING_FIELD_INDEX, authToken);
+    }
+
+    public static String[] getStringArrayField(BValue bValue) {
+        if (bValue != null && bValue instanceof BStringArray) {
```

Comment:
Is this done to get a new copy of the array, so the retrieve won't be able to change the source array values? .. This is because, this operation is relatively expensive compared to getting the internal array from BStringArray itself. We would have to add a new method to BStringArray to get the internal array though.

Reply:
No. This was done since BStringArray does not allow to get the internal array.